### ISSUE TITLE
metric_volatility.TrainSizeVolatility

### ISSUE THREAD

**Issue by timvink** [ing-bank/probatus#88]
**Problem Description**

Other than metric volatility coming from the seed and/or the data splits, an important driver for volatility is the size of the training data. In the drawing below, basically, instead of measuring the variance across three model variations, it's much more interesting to be aware of the gradient of model performance increase over data increase:

![image](https://user-images.githubusercontent.com/5570380/110316482-14639c00-800b-11eb-93f0-c18a068144db.png)

If the gradient is high, your model needs more data and your hence your model performance is not very 'stable'. If the gradient is low, it's more likely your model has learned what it can about your problem (`y`) from the data (`X`). 

The relevance of data size is confirmed by this [paper](https://arxiv.org/abs/2103.03098) (summarized in [this twitter thread](https://twitter.com/GaelVaroquaux/status/1367839174469029899?s=19)) & also shown by Vincent Warmerdam in his [PyData Cambridge 2021 meetup](https://www.youtube.com/watch?v=qcrR-Hd0LhI&t=388s) talk

**Desired Outcome**

Clear plot / report to assess if training data size might be a very important factor in metric uncertainty.

**Solution Outline**

Something like

```python
from sklearn.datasets import make_classification
from sklearn.ensemble import RandomForestClassifier
from probatus.metric_volatility import TrainSizeVolatility

X, y = make_classification(n_features=4)
clf = RandomForestClassifier()

volatility = TrainSizeVolatility(clf, iterations=10, repeats=10, test_prc = 0.3)
volatility_report = volatility.fit_compute(X, y)
volatility.plot()
```

To produce

![image](https://user-images.githubusercontent.com/5570380/110317471-7a045800-800c-11eb-8092-db22b66be502.png)

**Comment by sbjelogr** [comment]
Isn't this already covered by the learning curve in sklearn?
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.learning_curve.html

Or am I missing something?

@timvink When you talk about the gradient, you refer to the gradient of the performance wrt train data size?
This is actually a good check to do.

**Comment by Matgrb** [comment]
I really like the idea and i think it would fit pretty well into the theme of probatus.

However, as @sbjelogr indicated there is already very similar feature in sklearn, which might be even more effective because it uses CV. Would there be any value added of implementing this in probatus, other than adding plots?

**Comment by timvink** [comment]
> Isn't this already covered by the learning curve in sklearn?
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.learning_curve.html

🙈  You're right! Missed that one.

> Would there be any value added of implementing this in probatus, other than adding plots?

Providing a small wrapper for that class in `probatus`'s `.fit_compute()` and `.plot()` API could still be convenient. I'll leave it to you as maintainer to decide if you would like that feature or not; feel free to close this issue otherwise.

**Comment by sbjelogr** [comment]
I see the added value of computing the gradient (and probably even the hessian) of the learning curve, this is a nice idea, and not so obvious in the learning curve :)

**Comment by Matgrb** [comment]
I think it would be worth adding it to the package. The main use case behind it would be then to:
- Provide report returned by `compute`, that contains scores (mean and std) for different number of train samples, as well as the gradient and hessian for each of these. 
- Provide plots containing the relevant information, and convenience functions that allow to estimate how many samples one needs to training, in order to get a certain maximum std, or gradient.
- Adding documentation on how to analyse the data using such tool, and what kind of conclusions one may get.
I agree that the above points would be very useful in binary classification projects.

Implementation-wise, it is a relatively small feature, however, the issue would also include creating a documentation notebook and tests. One can split the issue into multiple PRs, e.g. 1st one with the code only, second one with tests, third one with tests, or simply implement only part of it.

While implementing this feature, keep in mind that [the method from sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.learning_curve.html) has many parameters. I would suggest the init and fit have similar parameters and default values as [ShapRFECV](https://github.com/ing-bank/probatus/blob/main/probatus/feature_elimination/feature_elimination.py), and allow for kwargs** that would be passed to the sklearn method.

Who would like to contribute to this feature? I would be happy to assist with discussions on implementation and the code reviews..

**Comment by Matgrb** [comment]
After looking further into this issue, I found the following[ page in sklearn docs](https://scikit-learn.org/stable/auto_examples/model_selection/plot_learning_curve.html#sphx-glr-auto-examples-model-selection-plot-learning-curve-py), which presents how to use the learning curve, and the code needed to make all the plots.

Seeing this, i suppose that the effort needed to implement that in probatus is incl docs, tests, etc exceeds the gains for the user. I will close this issue, if in the future we find more use cases for this feature, we can reopen it.
